x/logic/predicate/json.go

```diff
@@ -57,177 +131,185 @@
 //
 //	# JSON conversion to Prolog.
 //	- json_prolog('{"foo": "bar"}', json([foo=bar])).
-func JSONProlog(_ *engine.VM, j, p engine.Term, cont engine.Cont, env *engine.Env) *engine.Promise {
-	forwardConverter := func(in []engine.Term, _ engine.Term, env *engine.Env) ([]engine.Term, error) {
-		payload, err := prolog.TextTermToString(in[0], env)
-		if err != nil {
-			return nil, err
-		}
-
-		decoder := json.NewDecoder(strings.NewReader(payload))
-		term, err := decodeJSONToTerm(decoder, env)
+func JSONProlog(vm *engine.VM, j, p engine.Term, cont engine.Cont, env *engine.Env) *engine.Promise {
+	switch {
+	case prolog.IsGround(j, env):
+		payload, err := prolog.TextTermToString(j, env)
 		if err != nil {
-			return nil, err
-		}
-		if _, err := decoder.Token(); !errors.Is(err, io.EOF) {
-			return nil, engine.SyntaxError(AtomSyntaxErrorJSON.Apply(AtomMalformedJSON.Apply(engine.Integer(decoder.InputOffset()))), env)
+			return engine.Error(err)
 		}
+		is := engine.NewInputTextStream(strings.NewReader(payload))
+		defer is.Close()
 
-		return []engine.Term{term}, nil
-	}
-	backwardConverter := func(in []engine.Term, _ engine.Term, env *engine.Env) ([]engine.Term, error) {
+		return JSONRead(vm, is, p, cont, env)
+	default:
 		var buf bytes.Buffer
-		err := encodeTermToJSON(in[0], &buf, env)
-		if err != nil {
-			return nil, err
-		}
+		os := engine.NewOutputTextStream(&buf)
+		defer os.Close()
 
-		return []engine.Term{prolog.BytesToAtom(buf.Bytes())}, nil
+		return JSONWrite(vm, os, p, func(env *engine.Env) *engine.Promise {
+			return engine.Unify(vm, j, prolog.StringToAtom(buf.String()), cont, env)
+		}, env)
 	}
-	return prolog.UnifyFunctionalPredicate(
-		[]engine.Term{j}, []engine.Term{p}, prolog.AtomEmpty, forwardConverter, backwardConverter, cont, env)
 }
 
-func encodeTermToJSON(term engine.Term, buf *bytes.Buffer, env *engine.Env) (err error) {
+func encodeTermToJSON(term engine.Term, writer *textStreamWriter, env *engine.Env) (err error) {
 	switch t := term.(type) {
 	case engine.Atom:
 		if term == prolog.AtomEmptyList {
-			buf.Write([]byte("[]"))
+			if err := writeToStream(writer, []byte("[]"), env); err != nil {
+				return err
```

Comment:
Oh, yes, it would be better to have more coverage. However, this is white-box testing. You need to have a deep understanding of the action sequence to target a specific error at a certain depth. It seems particularly challenging, but if you’re up for it, I’d be curious to see how you approach it.

Reply:
Thanks for the help. Could you create a GitHub issue for this? We’ll address it in a second phase.